content/blog/overview-package-management-python.md

```diff
@@ -0,0 +1,363 @@
+for the sake of completeness.
+
+## pip-tools
+
+[pip-tools](https://github.com/jazzband/pip-tools) only tries to handle the package management part, meaning you still need to use
+`virtualenvwrapper` or something similar to handle virtualenvs.
+
+`pip-tools` gives you two command line tools: `pip-compile` and `pip-sync`.
+
+The easiest way to get started with `pip-tools` is to create a `requirements.in` file listing your dependencies:
+
+```txt
+boto3
+Django
+django-cors-headers
+django-filter
+django-floppyforms
+django-nested-inline
+django-storages
+djangorestframework
+djangorestframework-jsonp
+social-auth-app-django
+social-auth-core
+django-widget-tweaks
+```
+
+The `requirements.in` file uses the same requirements syntax as `requirements.txt`:
+I could indicate for example that I want `Django==2.0.8` instead of the latest version.
+
+Running `pip-compile` in the same folder will create a `requirements.txt` file:
+
+```txt
+#
+# This file is autogenerated by pip-compile
+# To update, run:
+#
+#    pip-compile --output-file requirements.txt requirements.in
+#
+boto3==1.8.7
+botocore==1.11.7          # via boto3, s3transfer
+certifi==2018.8.24        # via requests
+chardet==3.0.4            # via requests
+defusedxml==0.5.0         # via python3-openid, social-auth-core
+django-cors-headers==2.4.0
+django-filter==2.0.0
+django-floppyforms==1.7.0
+django-nested-inline==0.3.7
+django-storages==1.7
+django-widget-tweaks==1.4.2
+django==2.1.1
+
+... more lines
+```
+
+`pip-compile` automatically grabbed the latest versions and each line not present
+in your `requirements.in` has a comment indicating why it is here. This is already
+a big improvement over `pip freeze > requirements.txt`.
+
+Remember the packages hashes mentioned before? `pip-compile --generate-hashes` can automatically computes the hashes and put them
+in the generated file:
+
+```txt
+#
+# This file is autogenerated by pip-compile
+# To update, run:
+#
+#    pip-compile --generate-hashes --output-file requirements.txt requirements.in
+#
+boto3==1.8.7 \
+    --hash=sha256:7b54cc29dde1d4833b082b8ef4062872297cf652ed20b2d485e1cae544c7cdad \
+    --hash=sha256:b181fb87661a4268174ba29cb5efbc9e728202f3f80e00356e8135ce50a3799b
+botocore==1.11.7 \
+    --hash=sha256:4d36cb3a0b6308eeb56f69b964b92b9f4609423b07a4979c298dba694a5a4bd6 \
+    --hash=sha256:86d0bf23b5071c6a956ad2a1d5cd8d1d7c997e2c04151f7c8fff50779a8315d8 \
+    # via boto3, s3transfer
+certifi==2018.8.24 \
+    --hash=sha256:376690d6f16d32f9d1fe8932551d80b23e9d393a8578c5633a2ed39a64861638 \
+    --hash=sha256:456048c7e371c089d0a77a5212fb37a2c2dce1e24146e3b7e0261736aaeaa22a \
+    # via requests
+chardet==3.0.4 \
+    --hash=sha256:84ab92ed1c4d4f16916e05906b6b75a6c0fb5db821cc65e70cbd64a3e2a5eaae \
+    --hash=sha256:fc323ffcaeaed0e0a02bf4d117757b98aed530d9ed4531e3e15460124c106691 \
+    # via requests
+
+... more lines
+```
+
+The initial run with hashes took 16s and subsequent runs were below a second.
```

Comment:
16s to ... write a single txt file ? O_O

Reply:
To find the right set of dependencies, for `pip-tools` I didn't set pinned requirements. I should probably try with pinned one as well